1단계 - 엔티티 매핑 PR 요청드립니다.

earth-h opened the pull request:
안녕하세요, 리뷰어님 :) 

1단계 - 엔티티 매핑 PR 피드백 요청드립니다 !

- application.properties의 경우, yml을 주로 사용하고 있어 yml 형식으로 변경하여 올렸습니다.
- JPAAuditing을 위하여 BaseEntity를 구현하였습니다.
- 이미 가지고 있던 클래스들을 entity로 변경하는 작업인지라, git commit 메시지에 feat이 아닌 refactor로 나타내었습니다.
- 테스트마다, beforeEach를 통해 다른 테스트에서 저장한 row를 삭제하고자 하였습니다. 

감사합니다 :)

earth-h committed:
feat(properties): application.properties 삭제 및 application.yml 작성

application.yml
- Mysql 형식으로 쿼리를 보여주도록 설정
- H2 콘솔에 접근할 수도 있다고 생각해 콘솔 설정 진행
- application 실행 시마다 기존 테이블 삭제되고 생성되도록 ddl-auto를 create으로 설정

earth-h committed:
docs(guide): README.md 업데이트

JPA 1단계 - 엔티티 매핑 요구사항 정리 및 테이블 정의서 작성

earth-h committed:
fix(properties): application.yml 내 ddl-auto 위치 변경

application.yml
- spring.jpa.hibernate.ddl-auto: create 설정 (기존에 spring.jpa.properties.hibernate.ddl-auto로 잘못 넣음)

earth-h committed:
fix(properties): application.yml 내 오타 수정 및 필요하지 않은 설정 제거

application.yml
- spring.jpa.properties.hiberate -> spring.jpa.properties.hibernate로 오타 수정

earth-h committed:
docs(guide): README.md 업데이트

테스트 필요 목록 업데이트

earth-h committed:
feat(domain): BaseEntity 생성

BaseEntity
- JPA Auditing을 이용한 생성일, 마지막 수정시간 자동 저장을 위해 생성
  - @MappedSuperclass를 작성해 타 entity가 BaseEntity를 상속받을 때 BaseEntity 내 필드를 컬럼으로 인식하도록 함
  - createdAt의 경우, 생성 이후 변경되지 않도록 updatable = false로 설정

Application
- JPA Auditing 활성화를 위해 @EnableJpaAuditing 설정

earth-h committed:
refactor(domain, repository): Answer, AnswerRepository 리팩토링

Answer
- entity로 만들기 위해 @Entity 어노테이션 붙임
- @Entity 어노테이션은 public/protected 기본 생성자가 존재해야하므로, 다른 곳에서 접근하지 못하도록 protected로 선언

AnswerRepository
- 패키지 위치 변경

AnswerTest
- Answer 객체 생성 시, 예외가 발생할 수 있는 부분에 대한 테스트

AnswerRepositoryTest
- Answer 등록/조회/수정/삭제 테스트
- findByQuestionIdAndDeletedFalse() 메소드에 대해 테스트를 진행하지 않은 이유는, 아직 연관관계 매핑이 되어 있지 않아 questionId가 null이기 때문에 온전한 테스트를 할 수 없다고 생각하였기 때문

earth-h committed:
refactor(domain, repository): Question, QuestionRepository 리팩토링

Question
- entity로 만들기 위해 @Entity 어노테이션 붙임
- @Entity 어노테이션은 public/protected 기본 생성자가 존재해야하므로, 다른 곳에서 접근하지 못하도록 protected로 선언
- writeBy() 메소드 내 validation 추가

QuestionRepository
- 패키지 위치 변경

QuestionTest
- Question 객체 생성 시, 예외가 발생할 수 있는 부분에 대한 테스트

QuestionRepositoryTest
- Question 등록/조회/수정/삭제 테스트

earth-h committed:
refactor(domain, repository): User, UserRepository 리팩토링

User
- entity로 만들기 위해 @Entity 어노테이션 붙임
- @Entity 어노테이션은 public/protected 기본 생성자가 존재해야하므로, 다른 곳에서 접근하지 못하도록 protected로 선언
- Entity 내부에서만 사용되는 메소드 private으로 변경

UserRepository
- 패키지 위치 변경

UserRepositoryTest
- UserRepository 등록/조회/삭제 테스트
- User의 userId는 unique 제약조건이 걸려 있어 동일한 값 저장 가능 여부 테스트

earth-h committed:
refactor(domain, repository): DeleteHistory, DeleteHistoryRepository 리팩토링

DeleteHistory
- entity로 만들기 위해 @Entity 어노테이션 붙임
- @Entity 어노테이션은 public/protected 기본 생성자가 존재해야하므로, 다른 곳에서 접근하지 못하도록 protected로 선언
- id 조회용 getter 추가

DeleteHistoryRepository
- 패키지 위치 변경

DeleteHistoryRepositoryTest
- DeleteHistoryRepository 등록/조회/삭제 테스트

earth-h committed:
refactor(service): Repository 패키지 위치 변경으로 인한 import문 변경

DeleteHistoryService, QnaService, QnaServiceTest
- Repository 패키지 위치 변경으로 인해 import문 변경

mskangg commented on README.md:
```diff
@@ -0,0 +1,135 @@
+# JPA
+
+## 1단계 - 엔티티 매핑
+### 요구 사항
```
해야할 요구사항 먼저 정의 좋습니다 👍

mskangg commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,17 +1,31 @@
+    @Column(nullable = false)
     private boolean deleted = false;
 
+    protected Answer() {
+    }
```
Entity의 기본생성자를 `protected`로 하셨군요! 💯 
`private`으로 했을경우의 차이점에 대해 설명해주실수 있으실까용??

earth-h replied:
**[ public이 아닌 protected를 선택한 이유 ]**
`@Entity`를 사용하기 위해서는 protected, public 중 하나의 기본 생성자를 가지고 있어야 합니다.

domain 패키지 내부에 있는 entity들의 경우 대부분 `nullable = false`인 필드가 존재하기 때문에, 사실상 기본 생성자를 로직 내에서 사용하지 않을 것이라 판단해 public 보다는 접근 제한이 되는 protected로 구현하였습니다.

**[ no args constructor를 private로 했을 때와의 차이점 ]**
먼저, JPA가 entity 객체를 생성 방식은 어떠한 ORM 프레임워크를 사용하느냐에 따라 달라집니다. 많이 사용하고, 현 로직 구현 시 사용한 hibernate 같은 경우는 [hibernate문서](https://docs.jboss.org/hibernate/orm/5.4/quickstart/html_single/#hibernate-gsg-tutorial-basic-entity)에서 확인할 수 있듯이 Hibernate는 객체를 생성할 때에 Java Reflection을 사용하고 있어 no args 생성자가 필요합니다.

> The no-argument constructor, which is also a JavaBean convention, is a requirement for all persistent classes. Hibernate needs to create objects for you, using Java Reflection. The constructor can be private. However, package or public visibility is required for runtime proxy generation and efficient data retrieval without bytecode instrumentation.

Java Reflection 사용하는 것만으로는, no args 생성자를 private으로 할 수 없는 사유로 적합하지는 않습니다 :) 
다만, 위 docs에서 볼 수 있듯 **runtime proxy를 생성하고자 할 때에 package or public 가시성을 요구하기 때문에 private으로 사용할 수 없습**니다.

JPA를 사용하면, 연관관계를 맺은 entity에 대해서는 주로 LAZY 로딩을 하는 것으로 알고 있습니다. LAZY 로딩을 하면, 연관관계를 맺은 entity는 프록시 객체를 통해 가져오게 되고, 이 프록시 객체는 no args 생성자를 통해 만들어진 후 Java Reflection을 통해 필드값들을 넣게 되기 때문에 기본 생성자가 private일 수 없습니다.

**[ 참고한 자료 ]**
- [왜 JPA의 Entity는 기본 생성자를 가져야 하는가?](https://hyeonic.tistory.com/191)
- [stackoverflow](https://stackoverflow.com/questions/69082992/is-it-mandatory-to-use-no-args-constructor-using-entity-annotation-in-java-spri)

`@Entity`를 사용하기 위해서는 private 생성자가 불가능하다! 이렇게만 생각하고 넘어갔는데, 민수님 덕분에 왜일지 고민하는 시간을 가질 수 있었습니다! 👍🏻 

감사합니다 ㅎㅎ

mskangg replied:
> JPA를 사용하면, 연관관계를 맺은 entity에 대해서는 주로 LAZY 로딩을 하는 것으로 알고 있습니다. LAZY 로딩을 하면, 연관관계를 맺은 entity는 프록시 객체를 통해 가져오게 되고, 이 프록시 객체는 no args 생성자를 통해 만들어진 후 Java Reflection을 통해 필드값들을 넣게 되기 때문에 기본 생성자가 private일 수 없습니다.

hibernate는 lazy loading 시 cglib을 default로 사용하고 있는데, cglib은 상속을 통해 프록시객체를 만드는 방식입니다.
private 생성자만 존재한다면 상속을 위한 프록시 객체 생성시 부모객체를 초기화 할 수 없으니 최소 protected 로 선언해야 동작하는 부분도 알고 계시면 좋을 것 같아요 🙇

mskangg commented on src/main/java/qna/domain/BaseEntity.java:
```diff
@@ -0,0 +1,20 @@
+
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+public abstract class BaseEntity {
+
+    @CreatedDate @Column(nullable = false, updatable = false)
+    private LocalDateTime createdAt;
+    @LastModifiedDate
+    private LocalDateTime updatedAt;
```
공통 컬럼인 타임스탬프 좋아요 👍

mskangg commented on src/main/java/qna/repository/DeleteHistoryRepository.java:
```diff
@@ -1,6 +1,7 @@
-package qna.domain;
+package qna.repository;
```
repository라는 패키지를 새로 분리한 이유가 있으실까요??

earth-h replied:
repository용도와 domain용도를 분리하기 위함이었습니다! 편의에 의해 나눈 것이긴 합니다 ㅎㅎ

저는 주로 아래와 같이 용도를 분리하여 사용합니다 :) 
- domain
  - entity 관리
- repository
  - DB 접근과 관련된 항목
- service
- controller
- dto
  - entity가 아닌 일반 dto(예를 들어, API 호출 시 응답/요청을 관리할 패키지)

mskangg commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,17 @@
+        format_sql: true
+    hibernate:
+      ddl-auto: create-drop
+    show-sql: true
```
미션의 내용과는 별개지만 참고하시라고 코멘트 남깁니다.

![image](https://user-images.githubusercontent.com/27044096/199138439-4d8f7fce-1184-476c-9569-28d4a06aee44.png)

newline을 넣어주시면 좋을 것 같아요!
Github에서는 newline을 남기지 않으면 위와 같이 표식이 남습니다.

POSIX 표준에 맞지 않기도 하고 작업시 newline을 붙이는데, 다른 작업자가 붙이지 않을 경우 불필요한 file change가 생기기도 합니다.

저 같은 경우 IntelliJ에서 newline을 자동으로 추가하는 설정을 사용하고 있습니다.

![image](https://user-images.githubusercontent.com/27044096/199138693-c39d3d62-995a-4044-8848-4eb0393347a7.png)

`Ensure every saved file ends with a line break`

earth-h replied:
팁 감사합니다 !

말씀해주신 표식이 무엇인가 했었는데, newline이 없어서 발생한 것이었군요 ㅎㅎ 덕분에 원인을 알 수 있었습니다 👍🏻 

IntelliJ의 설정을 통해 마지막 newline 설정하는 부분은 꿀팁이네요! 팀에서 해당 부분 사용하지 않고 계신 분들께도 공유하면 좋을  것 같습니다!

mskangg commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,25 @@
     public static final Answer A2 = new Answer(UserTest.SANJIGI, QuestionTest.Q1, "Answers Contents2");
+
+    @Test
+    void Answer_객체_생성_시_Writer가_비어있으면_에러_발생() {
```
테스트명 작성에 좋은 글이 있어 링크 남겨드립니다.
정답은 없으니 지수님이 생각하시는대로 적용 부탁드려요!

https://blog.kingbbode.com/m/52?category=737337

mskangg commented on src/test/java/qna/repository/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Test
+    void 답변_조회_후_삭제하면_조회가_되지_않음_테스트() {
+        Answer saveAnswer = answerRepository.save(A2);
+        Long saveAnswerId = saveAnswer.getId();
+        Optional<Answer> findAnswer = answerRepository.findByIdAndDeletedFalse(saveAnswerId);
+
+        assertThat(findAnswer.isPresent()).isTrue();
+        findAnswer.ifPresent(answer -> answerRepository.delete(answer));
+        Optional<Answer> deletedAnswer = answerRepository.findByIdAndDeletedFalse(saveAnswerId);
+
+        assertThat(deletedAnswer.isPresent()).isFalse();
```
테스트 절 내엔 given, when, then 절이 명확한 것이 좋습니다!
이처럼 상태를 공유해야하는 테스트라면 `@DynamicTest`를 활용해보시는건 어떠실까요?

mskangg reviewed (changes requested):
안녕하세요, 지수님!
리뷰어를 맡게된 강민수입니다. 반갑습니다! 👐

이번 미션 깔끔하게 잘~ 구현하셨네요 ㅎㅎ
미션 수행전 jpa에 대해 어느정도 지식이 있어보이시는 것 같아 특별히 피드백 드릴 부분이 없어 다른부분에 몇가지 피드백 드렸어요!

고민해보시고 지수님 생각이 담긴 코드 반영해주세요! 🚀

earth-h committed:
refactor(properties): 마지막 newline 추가

POSIX 표준에 맞추기 위한 newline 추가

earth-h committed:
test(repository): UserRepositoryTest 리팩토링

UserRepositoryTest
- beforeEach 구문 내 deleteAllInBatch() 제거
  - @AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE) 설정을 하지 않았기 때문에 안심하고 날려도 되어서 넣었으나, 기본적으로 @Transactional이 설정된 @DataJpaTest를 사용했기 때문에 불필요한 설정이라 생각하여 제거
- 피드백을 통해 받은 DynamicTest를 상태 공유 필요 테스트에 적용
- 테스트명을 좀 더 명확하도록 수정

earth-h committed:
test(repository,domain): AnswerTest, AnswerRepositoryTest 리팩토링

AnswerRepositoryTest
- beforeEach 구문 내 deleteAllInBatch() 제거
- 피드백을 통해 받은 DynamicTest를 상태 공유 필요 테스트에 적용
- 테스트명을 좀 더 명확하게 수정

AnswerTest
- 테스트명을 좀 더 명확하게 수정

earth-h committed:
test(domain): AnswerTest 내 오타 수정

AnswerTest
- 오타수정

earth-h committed:
test(domain): QuestionTest

QuestionTest
- 테스트명을 좀 더 명확하게 수정

earth-h committed:
test(repository): DeleteHistoryRepositoryTest

DeleteHistoryRepositoryTest
- 테스트명을 좀 더 명확하게 수정
- beforeEach 구문 내 deleteAllInBatch() 제거
- 피드백을 통해 받은 DynamicTest를 상태 공유 필요 테스트에 적용

earth-h committed:
test(repository): UserRepositoryTest

UserRepositoryTest
- isPresent() 리팩토링

earth-h committed:
test(repository): QuestionRepositoryTest

QuestionRepositoryTest
- 테스트명을 좀 더 명확하게 수정
- beforeEach 구문 내 deleteAllInBatch() 제거
- 피드백을 통해 받은 DynamicTest를 상태 공유 필요 테스트에 적용

earth-h committed:
test(repository): AnswerRepositoryTest

AnswerRepositoryTest
- 테스트코드를 좀 더 직관적으로 변경

earth-h commented:
민수님 안녕하세요!

주신 피드백 보면서, 테스트코드를 좀 더 정리해보았습니다 :) 
- `DynamicTest`는 이번에 처음 알게된 터라 덕분에 새로운 테스트방식을 알게 되어 적용이 가능했습니다 👍🏻 
  - 다만, `DynamicTest`를 적용하니, 테스트코드가 길어지는데 문제가 없을지 피드백 부탁드립니다! (어플리케이션 로직만 메소드당 15줄을 넘지 않으면 되는것인지, 테스트코드 역시 15줄을 넘으면 안되는지 궁금합니다)
- 테스트메소드명을 좀 더 알아보기 쉽게 정리하였습니다. 혹, 좀 더 수정되었으면 좋겠다 싶은 부분이 있다면 피드백 주시면 고민해보겠습니다 :) 

감사합니다.

mskangg commented:
지수님! 피드백 반영한 부분이 보이지 않는데,, 혹시 커밋하고 푸시하신걸까요??

earth-h commented:
민수님..! 제가 푸시를 안 눌렀던 걸로 보입니다ㅠㅠ 푸시 눌러두었습니다..!

mskangg commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,21 @@
     public static final Question Q2 = new Question("title2", "contents2").writeBy(UserTest.SANJIGI);
+
+    @Test
+    void 질문_객체_생성_후_Writer_등록_할_때_Writer가_null이면_UnAuthorizedException_예외가_발생한다() {
```
이 부분은 개인마다 차이가 있을 수 있으니, 지수님이 생각하시는 방향대로 적용 부탁드려요!

`질문_객체_생성_후_Writer_등록_할_때_Writer가_null이면_UnAuthorizedException_예외가_발생한다.` -> `질문의 등록자가 없다면 예외를 발생시킨다.` 로 바꿀 수 있어보입니다!

개인적으로 `기능적 요구사항은 기획서에 표기된 요구사항이 그대로 옮겨질 수 있습니다.` 의 말처럼 요구사항을 정의한 이후 정의된 문장 그대로 테스트코드에 옮겨서 작성하곤 합니다.

이렇게 하면 기획 요구사항에 따른 테스트코드를 금방 유추할 수 있거니와 테스트코드 자체가 요구사항이 정리된 문서로 활용될 수 있습니다!

그렇기 때문에 테스트명에는 구체화된 클래스이름이나 변수명을 적지 않는 것이 좋습니다!
가령, 등록자라는 의미를 가진 `Writer`가 `Author`로 변경된다면 어떻게 될까요?
테스트코드 명을 전부 바꿔야하는 경우가 생깁니다.

그렇기에 요구사항 정의와 동시에 용어사전을 같이 정리하여 이 코드를 보는 모든 사람들이 `Question` -> `질문`, `Writer` -> `등록자` 와 같이 같은 문맥을 이해할 수 있게 정의하는 편이 가독성을 올리고 커뮤니케이션에 도움을 주곤 합니다.

earth-h replied:
`Question`, `Writer`와 같이 실제 명칭을 작성해야 더 이해하기 쉽지 않을까? 라는 생각에서 실제 필드명 또는 클래스 명으로 작성하였는데요 ! 말씀해주신 것처럼, 이후에 워딩이 변경될 수 있다는 부분을 놓쳤던 것 같습니다 :) 

어떤 예외가 발생하는지도, 테스트메소드명에 드러나야한다고 생각을 했는데, 이 역시 예외의 타입이 이후에 변경될 수 있으니, 명시하지 않는게 더 나을 수 있겠다는 생각이 드네요 ! 

주신 피드백 참고하여서 step2에서는 수정 진행해보겠습니다 :)

mskangg commented on src/test/java/qna/repository/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    @TestFactory
+    Collection<DynamicTest> 저장된_답변의_삭제여부에_따라_findByIdAndDeletedFalse메소드_조회_결과가_다르게_반환된다() {
```
상태 공유를 한다는 것은 시나리오 기반의 테스트일 확률이 높은데요!
`저장된_답변의_삭제여부에_따라_findByIdAndDeletedFalse메소드_조회_결과가_다르게_반환된다` -> `답변 삭제 시나리오` 와 같이 변경할 수 있어보입니다!

mskangg commented on src/test/java/qna/repository/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,89 @@
+        saveAnswer.setDeleted(false);
+        Long saveAnswerId = saveAnswer.getId();
+        return Arrays.asList(
+                DynamicTest.dynamicTest("저장한 답변은 findByIdAndDeletedFalse()로 조회하면 정상적으로 조회가 된다.", () -> {
```
세부적인 메서드명을 메세지로 치환해주세요!
메서드명이 변경된다면 해당 테스트코드명은 최신화가 안되어 추적이 불가능한 테스트코드가 되버릴 확률이 높습니다.

`저장한 답변은 findByIdAndDeletedFalse()로 조회하면 정상적으로 조회가 된다.` -> `삭제되지 않은 특정 답변을 조회한다`

`저장한 답변의 삭제여부를 true로 업데이트하면, 더 이상 findByIdAndDeletedFalse() 조회 시 조회되지 않는다.` -> `답변을 삭제하면 조회할 수 없다.`

테스트 명이 너무 길다면 불필요한 조건이나 설명이 들어가있을 확률이 높습니다.

하지만, 이러한 부분들은 같이 일하는 팀원분들과의 의견 조율이 필수적이니 이렇게 테스트명을 작성하기도 한다. 라고 봐주시면 좋을 것 같아요 🙇

earth-h replied:
고민했던 포인트가, `findById()`에 의한 조회인지, `findByIdAndDeletedFalse()`에 의한 조회인지 구분을 해야하지 않을까? 라는 부분이었습니다 !

`isDeleted()`와 `delete()`를 통해 실제 repository에서 삭제한게 다르다고 생각해서, 삭제여부와 삭제라는 단어를 다르게 사용하고 있는데요! 이로 인해 길어지는 것 같아 메소드명을 직접 넣었습니다.

그런데, 이전 피드백으로 주신 것처럼 필드명 변경에 의해 이 메소드 역시 언제든 변경될 수 있겠네요 !

mskangg reviewed (approved):
테스트코드의 가장 중요한 부분 중 하나는 같이 일하는 팀원들이 읽기 쉬운가? 라고 생각되어져요!
공통적으로 테스트의 의도가 잘 드러나도록 작성을 해야하지만, 테스트명 작성하는 부분도 중요하다고 생각해서 피드백 드렸습니다!

'저 팀에선 이렇게 작성하고 유지보수를 하고 있구나' 라고 생각해주시면 좋을 것 같아요!

이번 미션은 JPA 미션이니 이만 머지하고 다음 단계 때 지수님 생각이 담긴 피드백 반영 후 진행해주세요! 👍👍👍
